asf_public_discourse_home_decarbonisation/analysis/buildhub/extract_faqs.py

```diff
@@ -0,0 +1,245 @@
+                sentence = sentence.lower().replace(phrase, "")
+            sentences_without_inclusion.append(sentence)
+    return idk_phrases, sentences_without_inclusion
+
+
+if __name__ == "__main__":
+    # Read the CSV file
+    args = create_argparser()
+    category = args.category
+    collection_date_time = args.collection_date_time
+    # uses different getter function depending on the forum
+    if args.forum == "bh":
+        category_dataframe = get_bh_category_data(category, collection_date_time)
+    elif args.forum == "mse":
+        category_dataframe = get_mse_category_data(category, collection_date_time)
+    else:
+        logger.info("Please enter a valid forum name (bh or mse)")
+        sys.exit(-1)
+    FORUM_FAQ_PATH = os.path.join(
+        PROJECT_DIR, f"outputs/extracted_questions/{args.forum}/forum_{category}/"
+    )
+    # Ensure the output directory exists
+    os.makedirs(FORUM_FAQ_PATH, exist_ok=True)
+    # Tokenize sentences
+    nltk.download("punkt")
+    # Convert all text to strings to avoid type errors
+    category_dataframe["text"] = category_dataframe["text"].astype(str)
+    category_dataframe["sentences"] = category_dataframe["text"].apply(sent_tokenize)
+    # Apply the function to extract questions
+    category_dataframe["questions"] = category_dataframe["sentences"].apply(
+        extract_questions
+    )
+    (
+        category_dataframe["idk_phrases"],
+        category_dataframe["sentences_without_inclusion"],
+    ) = zip(*category_dataframe["sentences"].apply(extract_idk))
+    total_number_of_idk_phrases = category_dataframe["idk_phrases"].apply(len).sum()
+    total_number_of_sentences = category_dataframe["text"].apply(len).sum()
+    total_questions = category_dataframe["questions"].apply(len).sum()
+    # print some basic statistics!
+    logger.info(f"Total number of idk phrases: {total_number_of_idk_phrases}")
+    logger.info(f"Total number of questions: {total_questions}")
+    logger.info(f"Total number of sentences: {total_number_of_sentences}")
+    # Create a list of all questions
+    all_questions_in_category = sum(category_dataframe["questions"], [])
+    all_idk_phrases_in_category = sum(category_dataframe["idk_phrases"], [])
+    all_idk_phrases_without_inclusion_in_category = sum(
+        category_dataframe["sentences_without_inclusion"], []
+    )
+    # Create a new DataFrame with just the questions
+    logger.info("CSV files are located in: " + FORUM_FAQ_PATH + "\n")
+    questions_category_dataframe = pd.DataFrame(
+        all_questions_in_category, columns=["Question"]
+    )
+    # Check for duplicate questions
+    duplicate_questions = questions_category_dataframe[
+        questions_category_dataframe.duplicated(subset="Question")
+    ]
+    if not duplicate_questions.empty:
+        # output the duplicate questions
+        duplicate_questions.to_csv(
+            FORUM_FAQ_PATH + "duplicate_questions_" + category + ".csv", index=False
+        )
```

Comment:
This doesn't need to be done here - as it's already part of analysing the questions you're getting. You can just save the questions and then analyse the duplicates in a separate script.

In that separate script, I'd:

- load the questions df;
- apply .lower() to the "Question"
- then do df["Question"].value_counts()

This will give you the top questions.

Reply:
Yep this makes sense!